### ISSUE TITLE
Add a LightList<T> class

### ISSUE THREAD

**Issue by Patrick8639** [dotnet/runtime#47641]
Hello,
Often, the `List<T>` class is only used to save incoming data (from database or client application) only to keep them in a place and to enumerate on them, without the need to remove items or to sort them. In these cases, often, the number of items to keep is unknown.

There are two problems when using `List<T>`:
1. If the list contains more than approximately 10K (64-bit) or 20K (32-bit) elements, the internal array is allocated in the large object heap, making it difficult to reclaim the used memory.
2. When more items are added in the list, a new internal array must be allocated and the existing entries are copied to the new array.

The idea is to create a `LightList<T>` class, that implements `ICollection`and `IEnurable<T>` and can be used to add items and enumerate them.
The internal working should be the following:
1. Items are saved in linked arrays, so data is never copied.
2. When an internal array is full, a new array is added and is linked to the previous one. The size of the new array is the double of the size of the previous one, but limited to 8096 (to avoid the LOH).
3. When enumerating, the internal arrays are enumerated one after the other, in an iterator.

I have implemented this idea in a RemObjects Oxygene class, if there is interest to this source code, I will convert it to C# and publish it here. I've done some tests when reading 120000 items where I don't know the exact size before.
On my PC, the average execution time of the allocation is 61ms using `List<T>` and 51ms using `LightList<T>`, surely because I don't have the copying of 16 + 32 + 64 + 128 + 256 + 512 + 1024 + 2048 + 4096 + 8192 + 16384 " 32768 + 65535 items (total: about 917Kb). The timing for enumeration is about the same as of `List<T>`.
For memory use, `List<T>` has allocated blocks of 130Kb, 261Kb and 521Kb  in the LOH, so about 913K of memory that will surely never be reclaimed. The `LightList<T>` has never allocated more than 65536 bytes, so no memory is left in the LOH

**Comment by Clockwork-Muse** [comment]
> Often, the `List<T>` class is only used to save incoming data (from database or client application) only to keep them in a place and to enumerate on them, without the need to remove items or to sort them. In these cases, often, the number of items to keep is unknown.

... what do you mean by "keep" in this context?  Just that you don't know the number of items you expect to receive from the communication pipeline?

Note that if you're dealing with a database, you can probably avoid creating the list altogether by just enumerating over the result set:

```csharp
// Doing this:
foreach (var row in some_query):
{
    ProcessVar(var);
}

// instead of this
foreach (var in some_query.ToList()):
{
    ProcessVar(var);
}
```
... and that assumes you need to get all the rows into the application server in the first place, instead of running an `UPDATE` statement.  Also, databases are generally able to return an estimate of the rows to be returned from the statement - I'm wondering if the driver might be ignoring the value.

Depending on how client applications are contacting your server, you may be able to do something similar, and also potentially eliminate some LOH allocations you haven't mentioned (for instance, if your server is receiving JSON and you wait until the entire message is received to process it, you now have a string of roughly similar length sitting in the LOH).

> The size of the new array is the double of the size of the previous one, but limited to 8096 (to avoid the LOH).

The bigger thing you might be dodging is Gen 2 of the garbage collector, which would be far more dependent on how long-lived your list is.

> For memory use, `List<T>` has allocated blocks of 130Kb, 261Kb and 521Kb in the LOH, so about 913K of memory that will surely never be reclaimed.

If the program needs the memory, it will be reclaimed.  That it's not _currently_ being reclaimed is usually a feature; if your program runs for a short enough amount of time, or doesn't need that much memory, it's possible that the garbage collector never runs, among various optimizations.

**Comment by TonyValenti** [comment]
I definitely like this idea but I'd prefer to see these enhancements just made to the List class directly.

I often use a LinkedList to save results from a database to avoid the GC issues but I know that a linked list uses 3x more memory than a same sized list.

**Comment by danmosemsft** [comment]
A list backed by a series of arrays has also been proposed to enable very large lists https://github.com/dotnet/runtime/issues/46586

**Comment by Patrick8639** [comment]
@Clockwork-Muse 
In my case, more than 100 data types, To avoid having multiple copies of the code, there is a `GetAsync` generic method that takes a database connection and a filter, and that returns an `IEnumerable<T>` with the items. Dependeing on the data type and the filter, the result can be 0 item, 1, 100 or 100K. Big results come when we must to calculate statistics for a full year.
Some algarithms require multiple passes on the items, so a simple, read-only, enumerator is not always a solution.

> Also, databases are generally able to return an estimate of the rows to be returned from the statement - I'm wondering if the driver might be ignoring the value.

To do this, we need to make a `SELECT COUNT(*)`, which can take time for complex conditions (i.e. full-text search, or with deep joins for the conditions). The time price to just have the number of rows can be greater that the lost time used to reallocate.

> If the program needs the memory, it will be reclaimed. That it's not currently being reclaimed is usually a feature; if your program runs for a short enough amount of time, or doesn't need that much memory, it's possible that the garbage collector never runs, among various optimizations.

The list, in my case, is just used to process the data, which normally is fast but can be longer in some cases , so the allocated memory should remain in GEN 0. For the LOH, [this article](https://docs.microsoft.com/en-us/dotnet/standard/garbage-collection/large-object-heap), it is not often that the LOH is collected and it has a time penalty to copy the data on it. Allocating small objects is much more faster, on the GC side.

**Comment by Patrick8639** [comment]
@TonyValenti 
> I definitely like this idea but I'd prefer to see these enhancements just made to the List class directly.

The problem is that the `insert`, `Remove` and `RemoveAt` methods are more complicated to implement. And the `Sort`mthod would be a nightmare I think. So the idea to have a "light" list with limited functionalities.

> I often use a LinkedList to save results from a database to avoid the GC issues but I know that a linked list uses 3x more memory than a same sized list.
This is 3x the memory for the mechanism to keep the list linked, not for the data.
With the linked array solution, the memory increment is low (a simple pointer for 8K entries with big array.

**Comment by Clockwork-Muse** [comment]
> Dependeing on the data type and the filter, the result can be 0 item, 1, 100 or 100K. Big results come when we must to calculate statistics for a full year.

... I mean, if I anticipate I might be processing huge numbers of results, I'd be trying to run some portions in parallel, which would possibly also result in smaller arrays.  Although I'd likely try to run some or all of the processing in the database itself, to avoid transferring large chunks of data.

> Some algarithms require multiple passes on the items, so a simple, read-only, enumerator is not always a solution.

I don't know enough of what you're doing, but I'd be trying to do all the processing for an item individually.  That is, if you're running two aggregates, run the aggregates together:

```csharp
// instead of:
public void SomeProcessing()
{
    var data = connectedDatabase.SomeQuery().ToList();
    var aggregate1 = data.Where(filter1).Aggregate(aggregateOne);
    var aggregate2 = data.Where(filter2).Aggregate(aggregateTwo);
}
// doing this:
public void SomeProcessing()
{
    var data = connectedDatabase.SomeQuery();
    var (aggregate1, aggregate2) = data.Aggregate((item, cumulative) =>
        {
           // This looks a little messier, but there's ways to clean this up.
           (a, b) = cumulative;
            if (filter1(item)
           {
               a = aggregateOne(item, a);
           {
           if (filter2(item)
           {
               b = aggregateTwo(item, b);
           }
           return (a, b);
        });
}
```
> To do this, we need to make a `SELECT COUNT(*)`, which can take time for complex conditions (i.e. full-text search, or with deep joins for the conditions). The time price to just have the number of rows can be greater that the lost time used to reallocate.

Ah, no, you misunderstand.  When a database runs any statement it estimates or otherwise counts how many rows it thinks it will update or return (there's a number of reasons for this).  Depending on the query this value can be pretty spot on.  [This issue discusses one implementation of this](https://github.com/dotnet/runtime/issues/31533).  You shouldn't need to do this manually - I would have assumed that it was being used to do the initial allocation for the list.  Note that this won't prevent things being placed in the LOH, but would prevent most of the reallocations/copies.

**Comment by eiriktsarpalis** [comment]
Duplicate of #46586. I'm going to close this one.
